### PR TITLE
mount: allow / in archive names with rules that make sense

### LINKED ISSUE #2263: mount: allow / in archive names
Only seems to be forbidden because it would make the `borg mount` code a bit more complex. Parsing is not hard; extra assertion would probably be `posixpath.normpath(archive) == archive` (to rule `../../` & friends out) and it does not start with a /.

Allowing these gives a nicer, more intuitive separation of archives, e.g. `{hostname}/{date...}` is much neater with / as a separator than -_#@ (whichever) for mounting it, but also because / is considered a separator in general.

Btw., currently `.` and `..` are valid archive names; while these don't crash `mount`, they're inaccessible there.

### PR DESCRIPTION
**Pull request by enkore**
- POSIX_normalization(archive name) == archive name

  So foo/bar is valid,  foo/bar/, foo/./bar/, foo/../bar/ are not.

- May not begin with any number of slashes (i.e. not an absolute
  path)

- Archives may not be named "." or ".."

  This is a new restriction and previously broke borg-mount
  (it didn't crash, you just got EIO when reading the mountpoint dir)

  With this change borg-mount just ignores existing archives with
  these names.

- borg-mount is very reasonable. You can have an archive foo/bar,
  and an archive foo, if you mount the repository, you'll find
  foo/bar with the archive but also the contents of foo under
  foo/ (makes sense).

  I don't think this should be restricted. If someone does not
  want that behaviour, just don't name your archives like that.
  I think it can be even quite useful.

  Another question might be, "if I have an archive foo, that
  has a directory bar, but also an archive foo/bar, which one
  will I get?". Answer: --sort-by decides, first come first served.

  Again, this is something that doesn't need restricting. Don't
  want that behaviour (which again can be quite useful) -> don't
  name your archives that way.

Fixes #2263

### CODE REVIEW

File: docs/usage_general.rst.inc [hunk 1]
```diff
@@ -60,10 +60,28 @@ Repository / Archive Locations
...
-shell or in a filesystem (borg mount will use the archive name as directory
-name).
+Borg does not assume a particular meaning or structuring of archive names,
+but for use with :ref:`borg_prune` and other commands it makes sense
+to structure names with prefixes (e.g. ``{hostname}-{date}``) for the
+``--prefix`` option.
+
+When mounting an entire repository with :ref:`borg_mount` (without the
+'versions' option) the archive names are used as paths in the
+virtual filesystem.
```
Comment (ThomasWaldmann) [inline docs/usage_general.rst.inc]: well, this somehow contradicts the first paragraph.
after this proposed change, borg mount assumes that the archive name is a path (without a slash it is still a 1-level path).

File: src/borg/helpers.py [hunk 2]
```diff
@@ -1061,7 +1062,7 @@ class Location:
     optional_archive_re = r"""
         (?:
             ::                                              # "::" as separator
-            (?P<archive>[^/]+)                              # archive name must not contain "/"
+            (?P<archive>[^/]{1}.*)
```
Comment (ThomasWaldmann) [inline src/borg/helpers.py]: hmm, what is the point of having `{1}` there?

File: src/borg/helpers.py [hunk 3]
```diff
@@ -29,6 +29,7 @@
 from operator import attrgetter
 from string import Formatter
 from shutil import get_terminal_size
+from posixpath import normpath
```
Comment (ThomasWaldmann) [inline src/borg/helpers.py]: rather use import posixpath here to make it clearer below that normpath is from there and not from os.path?

**Comment by ThomasWaldmann** [comment]
pr title should reflect this is a change to how borg mount works.

**Comment by enkore** [comment]
merge?

**Comment by ThomasWaldmann** [comment]
don't merge yet, see discussion in the ticket.

**Comment by enkore** [comment]
I thought we were done with that discussion

**Comment by ThomasWaldmann** [comment]
@textshell ^^

File: docs/usage_general.rst.inc [hunk 4]
```diff
@@ -60,10 +60,25 @@ Repository / Archive Locations
...
-name).
+Archive names can be almost any string, but Borg places two important restrictions
+on them:
+
+1. Since archives can only be filtered by ``--prefix`` (for e.g. :ref:`borg_prune`),
+   grouping needs to be prefix based. For example, when backing up multiple hosts
+   into the same repository ``{hostname}-{date}`` makes sense, while ``{date}-{hostname}``
+   will pose issues for pruning.
+2. When mounting (:ref:`borg_mount`) an entire repository, archive names are interpreted
+   as paths. Therefore archives cannot be named ".", "..", cannot start nor end with a slash,
```
Comment (ThomasWaldmann) [inline docs/usage_general.rst.inc]: ... may not end with "." or ".." and can not contain "/./" or "/../", ...

File: docs/usage_general.rst.inc [hunk 5]
```diff
@@ -60,10 +60,25 @@ Repository / Archive Locations
...
+   and may not contain two or more consecutive slashes.
+
+Otherwise any string can be used, including whitespace, newlines, Unicode
+and so on, however, to cause less confusion and maximize portability it's
+likely a good idea to avoid special characters. Similarly backslashes are permitted,
+but should not be used to improve portability.
```
Comment (ThomasWaldmann) [inline docs/usage_general.rst.inc]: Could we just disallow backslashes?

File: src/borg/helpers.py [hunk 6]
```diff
@@ -1178,6 +1177,10 @@ def normpath_special(p):
             return True
         return False
 
+    @staticmethod
+    def archive_valid(archive):
+        return not archive or (archive == posixpath.normpath(archive) and archive not in ('..', '.') and '::' not in archive)
```
Comment (ThomasWaldmann) [inline src/borg/helpers.py]: handle the "not archive" in the caller, not here.

File: src/borg/helpers.py [hunk 7]
```diff
@@ -1230,8 +1233,8 @@ def validator(text):
 
 def archivename_validator():
     def validator(text):
-        if '/' in text or '::' in text or not text:
-            raise argparse.ArgumentTypeError('Invalid repository name: "%s"' % text)
+        if not Location.archive_valid(text):
```
Comment (ThomasWaldmann) [inline src/borg/helpers.py]: archive_valid() currently accepts empty archive name, which was explicitly forbidden before, see 2 lines above.

File: docs/usage_general.rst.inc (hunk 5 above)
Comment (enkore) [inline docs/usage_general.rst.inc]: #2290.

File: src/borg/helpers.py (hunk 6 above)
Comment (enkore) [inline src/borg/helpers.py]: https://docs.python.org/3/library/os.path.html#os.path.normpath

> However, you can also import and use the individual modules if you want to manipulate a path that is always in one of the different formats. They all have the same interface: posixpath for UNIX-style paths

File: src/borg/fuse.py [hunk 8]
```diff
@@ -138,6 +137,23 @@ def _create_dir(self, parent):
...
 
+    def _create_dir_parents(self, path, parent=1):
+        """Create directory tree (equivalent to mkdir --parents)."""
+        segments = path.split(b'/')
+        segments.reverse()
+        inode = parent
+        while segments:
+            try:
+                inode = self.contents[inode][segments[-1]]
+            except KeyError:
```
Comment (enkore) [inline src/borg/fuse.py]: NB. specifically this is a common pattern in the entire FUSE code via `try: _find_inode except KeyError:`.

Comment (enkore) [inline src/borg/fuse.py]: > What is the right response here if self.contents[inode] throws?

raise Exception, _exit(2).

File: src/borg/helpers.py (hunk 6 above)
Comment (ThomasWaldmann) [inline src/borg/helpers.py]: what I meant is that a falsy archive name is not a valid archive name.

the only case where this is used in a overall-valid context is if we have a repo location without an archive name, but that should be handled outside of an archive_valid function.

**Comment by ThomasWaldmann** [comment]
IIRC the outcome of the discussion was "not wanted without changes" not "not wanted".

**Pull request closed without merging**